### PR TITLE
guest-provider: Error if used outside of development environment 

### LINKED ISSUE #24173: 🐛 Bug Report: Warns when guest provider is explicitly disabled
### 📜 Description

With the new backend system, `auth-backend-module-guest-provider` warns about initialising the guest provider here 

https://github.com/backstage/backstage/blob/eec9a51d5e04fffae336d4c231a6acc79b051b8b/plugins/auth-backend-module-guest-provider/src/module.ts#L41

even though `auth.providers.guest` is set explicitly to `null`

### 👍 Expected behavior

There should be no logging when explicitly disabled.

### 👎 Actual Behavior with Screenshots

The log message `You should NOT be using the guest provider outside of a development environment.` was outputted

### 👟 Reproduction steps

Include the following in `packages/backend/index.ts`:

```ts
backend.add(import('@backstage/plugin-auth-backend-module-guest-provider'));
```

And the following in your `app-config.yaml`

```yaml
auth:
  providers:
    guest: null
```

Run the server with `NODE_ENV=production`

### 📃 Provide the context for the Bug.

_No response_

### 🖥️ Your Environment

_No response_

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by nikolaik**
Spare us one warning when running with `NODE_ENV !== 'development'`.

Fixes #24173

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

File: plugins/auth-backend-module-guest-provider/src/module.ts [hunk 1]
```diff
@@ -36,6 +36,11 @@ export const authModuleGuestProvider = createBackendModule({
         config: coreServices.rootConfig,
       },
       async init({ providers, logger, config }) {
+        const providerConfig = config.getOptionalConfig('auth.providers.guest');
+        if (!providerConfig) {
+          return;
+        }
```
Comment (freben) [inline plugins/auth-backend-module-guest-provider/src/module.ts:42]: Alright, we talked about this one among the maintainers and we agree that the initial log is just plain wrong. Would you mind to update this PR to do things a bit differently, making it better for everyone? Would appreciate it 🙏 

Summary:

- Remove this added code, and also remove the `if` and logging below. Essentially leave _only_ the `providers.registerProvider` line in here
- Modify [the initialize function of the authenticator](https://github.com/backstage/backstage/blob/e9180e8b757e295cb11325ae0387e07fc1408a33/plugins/auth-backend-module-guest-provider/src/authenticator.ts#L23) instead
  - change it to `initialize({ config })` - it gets that passed in
  - make it check whether it's outside of the development env, AND `dangerouslyAllowOutsideDevelopment` isn't set to true
  - if so, throw a hard error similar to the one below, like `The guest provider cannot be used outside of a development environment`, which rejects the login

Comment (nikolaik) [reply plugins/auth-backend-module-guest-provider/src/module.ts:42]: I do like making thing better for everyone!🤗 I'll update my PR as suggested 😊

Comment (nikolaik) [reply plugins/auth-backend-module-guest-provider/src/module.ts]: Updated with your suggestions <3

File: .changeset/purple-waves-smile.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-auth-backend-module-guest-provider': patch
+---
+
+Error if used outside of a development environment
```
Comment (pjungermann) [inline .changeset/purple-waves-smile.md:5]:
```suggestion
Error if used outside of a development environment without explicit allowance
```

File: plugins/auth-backend-module-guest-provider/src/authenticator.ts [hunk 3]
```diff
@@ -15,12 +15,22 @@
...
+import { NotImplementedError } from '@backstage/errors';
 
 export const guestAuthenticator = createProxyAuthenticator({
   defaultProfileTransform: async () => {
     return { profile: {} };
   },
-  initialize() {},
+  initialize({ config }) {
+    const allowOutsideDev = config.getOptionalBoolean(
+      'backend.auth.dangerouslyAllowOutsideDevelopment',
```
Comment (Rugvip) [inline plugins/auth-backend-module-guest-provider/src/authenticator.ts:26]: This config keys lives in `auth.providers.guest`, but this config object as already scoped to that so you can skip the prefix altogether. We don't necessarily need a test here, but please verify that it's working as intended manually.

Comment (nikolaik) [reply plugins/auth-backend-module-guest-provider/src/authenticator.ts:26]: Of cours, thank you for explaing. I'll fix

Comment (nikolaik) [reply plugins/auth-backend-module-guest-provider/src/authenticator.ts:26]: The error message running with `NODE_ENV=production`: 
```
ForwardedError: Plugin 'auth' startup failed; caused by Error: Failed to initialize guest auth provider, The guest provider cannot be used outside of a development environment
```

**Review by freben — APPROVED** [review]
Nice!

**Pull request merged**